### PR TITLE
feat(smartmodules): added `json-sql` SmartModule

### PR DESCRIPTION
**Pull request by galibey**
The SmartModule transforms the data in arbitrary JSON format into the defined SQL model. The transformation is defined by the mapping, which is JSON as well and is passed to the module during `init` phase. 

Related #328

### CODE REVIEW

File: rust-connectors/models/fluvio-model-sql/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+[package]
+name = "fluvio-model-sql"
+version = "0.0.1"
+authors = ["Fluvio Contributors <team@fluvio.io>"]
+edition = "2021"
+publish = false
```
Comment (sehz) [inline rust-connectors/models/fluvio-model-sql/Cargo.toml:6]: add license

File: smartmodules/json-sql/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,19 @@
+[package]
+name = "json-sql"
+version = "0.0.0"
+authors = ["Fluvio Contributors <team@fluvio.io>"]
+description = "JSON to SQL transformation SmartModule"
+edition = "2021"
+publish = false
```
Comment (sehz) [inline smartmodules/json-sql/Cargo.toml:7]: add license

**Review by sehz — CHANGES_REQUESTED** [review]
LGTM.  Couple of minor nits

File: rust-connectors/models/fluvio-model-sql/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+pub enum Operation {
+    Insert { table: String, values: Vec<Value> },
+}
+
+/// Value with SQL column name and supported SQL type.
+#[derive(Debug, Serialize, Deserialize, Eq, PartialEq)]
+pub struct Value {
+    pub column: String,
+    pub raw_value: String,
+    pub r#type: Type,
```
Comment (simlay) [inline rust-connectors/models/fluvio-model-sql/src/lib.rs:15]: There's nothing wrong with using `r#type` as a field or varable name. I personally prefer to use `type_` with [`#[serde(rename = "type"))]`](https://serde.rs/field-attrs.html#rename). It just makes it easier to _type_.

File: smartmodules/json-sql/src/mapping.rs [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+#[derive(Debug, Serialize, Deserialize, Eq, PartialEq)]
+pub struct Column {
+    #[serde(alias = "json-key")]
+    pub json_key: String,
+    pub value: Value,
+}
+
+#[derive(Debug, Serialize, Deserialize, Eq, PartialEq)]
+pub struct Value {
+    pub r#type: ValueType,
```
Comment (simlay) [inline smartmodules/json-sql/src/mapping.rs:24]: https://github.com/infinyon/fluvio-connectors/pull/329/files#r959796018

**Review by simlay — COMMENTED** [review]
Other than a few nits (I like `type_` over `r#type`). I think this looks good.

File: rust-connectors/models/fluvio-model-sql/Cargo.toml (hunk 1 above)
Comment (galibey) [reply rust-connectors/models/fluvio-model-sql/Cargo.toml:7]: Done

File: smartmodules/json-sql/Cargo.toml (hunk 2 above)
Comment (galibey) [reply smartmodules/json-sql/Cargo.toml:8]: Done

File: rust-connectors/models/fluvio-model-sql/src/lib.rs (hunk 3 above)
Comment (galibey) [reply rust-connectors/models/fluvio-model-sql/src/lib.rs]: Yeah, I agree. Done.

File: smartmodules/json-sql/src/mapping.rs (hunk 4 above)
Comment (galibey) [reply smartmodules/json-sql/src/mapping.rs]: Done

File: smartmodules/json-sql/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+[lib]
+crate-type = ['cdylib', 'rlib']
+
+[dependencies]
+fluvio-model-sql = { path = "../../rust-connectors/models/fluvio-model-sql" }
+fluvio-smartmodule = { git = "https://github.com/infinyon/fluvio" }
```
[the remark this answers is not in the document]
Comment (galibey) [reply smartmodules/json-sql/Cargo.toml:15]: Unfortunately, I didn't manage to do it here. Due to different versions between the patch and the original (0.4.0 vs 0.2.5) and different features set, cargo complains and says that the patch is not applied.

[the remark this answers is not in the document]
Comment (galibey) [reply smartmodules/json-sql/Cargo.toml]: Careful reading documentation helped :).  Fixed.

File: rust-connectors/models/fluvio-model-sql/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,13 @@
+[package]
+name = "fluvio-model-sql"
+version = "0.0.1"
+authors = ["Fluvio Contributors <team@fluvio.io>"]
+edition = "2021"
+license = "Apache-2.0"
+publish = false
+description = "The data model intended for using inside SQL connector"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
```
Comment (sehz) [inline rust-connectors/models/fluvio-model-sql/Cargo.toml:10]: don't think we need this?

**Review by sehz — APPROVED** [review]
LGTMS.  One minor nit but that can be done in cleanup PR

File: Cargo.toml [hunk 7]
```diff
@@ -11,11 +11,16 @@ members = [
...
     "rust-connectors/utils/test-connector",
     "rust-connectors/utils/fluvio-smartstream-map/",
+    "smartmodules/json-sql",
     "examples/github-stars/github-stars-smartmodule-map/",
     "examples/github-stars/slack-display-smartmodule-map/",
 ]
 resolver = "2"
+
+[patch.crates-io]
+fluvio-smartmodule = { git = "https://github.com/infinyon/fluvio" }
```
Comment (simlay) [inline Cargo.toml:26]: It'd probably be good to have the `rev` pointing to the exact commit so it won't break as easily. This has been sitting in review long enough that I don't feel strongly about it.

**Review by simlay — APPROVED** [review]
LGTM.

File: rust-connectors/models/fluvio-model-sql/Cargo.toml (hunk 6 above)
Comment (galibey) [reply rust-connectors/models/fluvio-model-sql/Cargo.toml]: Done

File: Cargo.toml (hunk 7 above)
Comment (galibey) [reply Cargo.toml]: Very good point. Thanks. Done.

**Review by simlay — APPROVED** [review]

**Pull request merged**